### PR TITLE
Add eslint rule for not allowing styled-jsx in _document.js

### LINKED ISSUE #32656: ESLint rule idea: `<style>` is not allowed inside _document.js
### Describe the feature you'd like to request

As per the docs: https://nextjs.org/docs/advanced-features/custom-document#caveats
styled-jsx is not allowed inside custom `_document.js` and I was getting 500 errors on all SSR pages when deploying a site to production that had a `<style>` tag inside a custom `_document.js`.

I think it's a good idea to add an ESLint rule to remind users about this since the error I got did not make it straightforward to figure out the root cause of the issue.

### Describe the solution you'd like

Add a new ESLint rule warning about the use of `<style>` tags inside a custom `_document.js`

### Describe alternatives you've considered

Maybe prevent the build if a `<style>` tag is detected inside a custom `_document.js`

### PR DESCRIPTION
**Pull request by src200**
## Bug

- [ ] Related issues linked
fixes #32656

### CODE REVIEW

**Comment by src200** [comment]
Hey @timneutkens 
This is my first ever PR to my favorite framework(Actually this is my first opensource PR). Please let me know if I'm on the correct path.

**Review by balazsorban44 — COMMENTED** [review]
It's not all style that is disallowed, only solutions like `styled-jsx`

https://nextjs.org/docs/advanced-features/custom-document#caveats

**Comment by src200** [comment]
> It's not all style that is disallowed, only solutions like `styled-jsx`
> 
> https://nextjs.org/docs/advanced-features/custom-document#caveats

Ah, gotcha! Updating it now.
Thanks

File: packages/eslint-plugin-next/lib/index.js [hunk 1]
```diff
@@ -38,6 +39,7 @@ module.exports = {
         '@next/next/no-document-import-in-page': 2,
         '@next/next/no-head-import-in-document': 2,
         '@next/next/no-script-in-document': 2,
+        '@next/next/no-style-in-document': 2,
```
Comment (balazsorban44) [inline packages/eslint-plugin-next/lib/index.js:42]: The name should probably be different, as mentioned, not every type of `<style>` is disallowed. This name change should be reflected in the file name, and the rule description etc. as well.

File: test/unit/eslint-plugin-next/no-style-in-document.test.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    ecmaFeatures: {
+      modules: true,
+      jsx: true,
+    },
+  },
+})
+const ruleTester = new RuleTester()
+
+ruleTester.run('no-style-in-document', rule, {
+  valid: [
```
Comment (balazsorban44) [inline test/unit/eslint-plugin-next/no-style-in-document.test.ts:16]: Could this add a regular `<style>` tag as that should be allowed?

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Added some minor comments.

File: packages/eslint-plugin-next/lib/index.js (hunk 1 above)
Comment (src200) [reply packages/eslint-plugin-next/lib/index.js:42]: Updated rule name and file names as well

File: test/unit/eslint-plugin-next/no-style-in-document.test.ts (hunk 2 above)
Comment (src200) [reply test/unit/eslint-plugin-next/no-style-in-document.test.ts]: Yes. So basically the rule only checks for `<style>` with `jsx` attribute.

Comment (src200) [reply test/unit/eslint-plugin-next/no-style-in-document.test.ts]: I have added another valid testcase.

File: packages/eslint-plugin-next/lib/index.js (hunk 1 above)
Comment (balazsorban44) [reply packages/eslint-plugin-next/lib/index.js:42]: only this actual line was forgotten to be renamed 😄

Comment (src200) [reply packages/eslint-plugin-next/lib/index.js]: > only this actual line was forgotten to be renamed 😄

Sorry! renamed.

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
ESLint errors also link to a document page that explains the issue in detail, and how the user could fix them. See any other lint rule as an example.

Document page example: https://github.com/vercel/next.js/blob/canary/errors/inline-script-id.md

**Comment by src200** [comment]
> ESLint errors also link to a document page that explains the issue in detail, and how the user could fix them. See any other lint rule as an example.
> 
> Document page example: https://github.com/vercel/next.js/blob/canary/errors/inline-script-id.md

@balazsorban44 Added documentation

File: errors/no-styled-jsx-in-document.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
+# No styled-jsx in Document
+
+### Why This Error Occurred
+
+Custom css like `styled-jsx` is not allowed in `pages/_document.{js,tsx}`.
```
Comment (balazsorban44) [inline errors/no-styled-jsx-in-document.md:5]:
```suggestion
Custom CSS like `styled-jsx` is not allowed in a Custom Document.
```

File: errors/no-styled-jsx-in-document.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+### Possible Ways to Fix It
+
+If you need share css for all your pages, take a look at the `pages/_app.{js,tsx}` or define a custom layout.
```
Comment (balazsorban44) [inline errors/no-styled-jsx-in-document.md:9]:
```suggestion
If you need shared CSS for all of your pages, take a look at the [Custom App](https://nextjs.org/docs/advanced-features/custom-app) file or define a custom layout.
```

File: errors/no-styled-jsx-in-document.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+}
+```
+
+These styles (`styles.css`) will apply to all pages and components in your application.
+
+### Useful links
+
+- [Custom Document Caveats](https://nextjs.org/docs/advanced-features/custom-document#caveats)
+- [Layouts](https://nextjs.org/docs/basic-features/layouts)
+- [Built in CSS Support](https://nextjs.org/docs/basic-features/built-in-css-support)
```
Comment (balazsorban44) [inline errors/no-styled-jsx-in-document.md:40]:
```suggestion
- [Built in CSS Support](https://nextjs.org/docs/basic-features/built-in-css-support)
- [Custom `App`](https://nextjs.org/docs/advanced-features/custom-app)
```

**Review by balazsorban44 — DISMISSED** [review]
Thanks, left some minor formatting-related comments and added some useful links.

**Comment by src200** [comment]
> Thanks, left some minor formatting-related comments and added some useful links.

Updated docs.
Thanks

**Comment by src200** [comment]
@balazsorban44 Could you please review this PR?

**Comment by ijjk** [comment]
## Failing test suites 
Commit: d49f2783c8fbb99f2b9ab48f0f2020792c2d3303

`yarn testheadless test/integration/amphtml/test/index.test.js` 
- AMP Usage > AMP dev mode > should detect the changes and display it

<details>
<summary>Expand output</summary>

  ● AMP Usage › AMP dev mode › should detect the changes and display it

    TIMED OUT: /This is the hot AMP page/

    This is a cold AMP page.

    no AMP for you...

      498 |
      499 |   if (hardError) {
    > 500 |     throw new Error('TIMED OUT: ' + regex + '\n\n' + content)
          |           ^
      501 |   }
      502 |   return false
      503 | }

      at Object.check (lib/next-test-utils.js:500:11)
      at Object.<anonymous> (integration/amphtml/test/index.test.js:363:9)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | src200/next.js no-style-in-document | Change |
| - | - | - | - |
| buildDuration | 14.5s | 15s | ⚠️ +449ms |
| buildDurationCached | 5.9s | 6.1s | ⚠️ +210ms |
| nodeModulesSize | 479 MB | 479 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | src200/next.js no-style-in-document | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.64 | 3.569 | -0.07 |
| / avg req/sec | 686.76 | 700.38 | +13.62 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.257 | 1.213 | -0.04 |
| /error-in-render avg req/sec | 1989.08 | 2060.46 | +71.38 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | src200/next.js no-style-in-document | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 29.2 kB | 29.2 kB | ✓ |
| webpack-HASH.js gzip | 1.54 kB | 1.54 kB | ✓ |
| Overall change | 72.9 kB | 72.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | src200/next.js no-style-in-document | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | src200/next.js no-style-in-document | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 193 B | 193 B | ✓ |
| amp-HASH.js gzip | 308 B | 308 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.71 kB | 2.71 kB | ✓ |
| head-HASH.js gzip | 359 B | 359 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.73 kB | 5.73 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.65 kB | 2.65 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 391 B | 391 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16 kB | 16 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | src200/next.js no-style-in-document | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | src200/next.js no-style-in-document | Change |
| - | - | - | - |
| index.html gzip | 533 B | 533 B | ✓ |
| link.html gzip | 547 B | 547 B | ✓ |
| withRouter.html gzip | 528 B | 528 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | src200/next.js no-style-in-document | Change |
| - | - | - | - |
| buildDuration | 16.5s | 16.6s | ⚠️ +84ms |
| buildDurationCached | 6s | 6.1s | ⚠️ +18ms |
| nodeModulesSize | 479 MB | 479 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | src200/next.js no-style-in-document | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.523 | 3.551 | ⚠️ +0.03 |
| / avg req/sec | 709.72 | 703.94 | ⚠️ -5.78 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.209 | 1.295 | ⚠️ +0.09 |
| /error-in-render avg req/sec | 2067.69 | 1930.99 | ⚠️ -136.7 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | src200/next.js no-style-in-document | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.7 kB | 42.7 kB | ✓ |
| main-HASH.js gzip | 29.6 kB | 29.6 kB | ✓ |
| webpack-HASH.js gzip | 1.54 kB | 1.54 kB | ✓ |
| Overall change | 74 kB | 74 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | src200/next.js no-style-in-document | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | src200/next.js no-style-in-document | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 311 B | 311 B | ✓ |
| css-HASH.js gzip | 324 B | 324 B | ✓ |
| dynamic-HASH.js gzip | 2.89 kB | 2.89 kB | ✓ |
| head-HASH.js gzip | 357 B | 357 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.82 kB | 5.82 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.78 kB | 2.78 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 392 B | 392 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.3 kB | 16.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | src200/next.js no-style-in-document | Change |
| - | - | - | - |
| _buildManifest.js gzip | 457 B | 457 B | ✓ |
| Overall change | 457 B | 457 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | src200/next.js no-style-in-document | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 547 B | 547 B | ✓ |
| withRouter.html gzip | 529 B | 529 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

</details>
Commit: d49f2783c8fbb99f2b9ab48f0f2020792c2d3303

**Review by ijjk — APPROVED** [review]
Thanks for the PR!

**Pull request merged**